File: GullyTrends/ViewControllers/FilterViewController/FilterViewController.m
//
//  FilterViewController.m
//  Constraintsdemo
//
//  Created by Lakshmi Vajrapu on 07/12/19.
//  Copyright © 2019 Lakshmi Vajrapu. All rights reserved.
//

#import "FilterViewController.h"

@interface FilterViewController ()
@property (strong, nonatomic) IBOutlet UITableView *leftTableView;
@property (strong, nonatomic) IBOutlet UITableView *rightTableview;
@property (strong, nonatomic) NSArray *leftdataArray;
@property (strong, nonatomic) NSArray *rightdataArray;
@property (assign) NSInteger selectedleftindex;
@property (strong) NSMutableArray *selectedrighttindexs;

@end

@implementation FilterViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.leftTableView.delegate=self;
    self.rightTableview.delegate=self;
    
    self.leftdataArray=@[@"Color",@"Material",@"Size",@"Sorting",@"Pattern",@"Type",@"Brand"];
    self.rightdataArray =@[@"Color",@"Material",@"Size",@"Sorting",@"Pattern",@"Type",@"Brand"];
   
    self.selectedrighttindexs = [[NSMutableArray alloc]init];
    [self.rightTableview setEditing:YES animated:YES];
    self.leftTableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];

}


# pragma UITableview Delegate Methods

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    if (tableView == self.leftTableView) {
        return self.leftdataArray.count;
    }else{
        return self.rightdataArray.count+10;
    }
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return 50;
}

-(UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSIndexPath*)indexPath
{
    if (tableView == self.leftTableView)
    {
        static NSString *CellIdentifier1 = @"cellForleftTableView";

        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier1];

        if (!cell)
        {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier1];
        }

        cell.textLabel.text = [NSString stringWithFormat:@"%@",[self.leftdataArray objectAtIndex:indexPath.row]];

        return cell;
    }
    else
    {
        static NSString *CellIdentifier2 = @"cellForrightTableview";

        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier2];

        if (!cell)
        {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier2];
        }

        cell.textLabel.text = [NSString stringWithFormat:@"%@",[self.rightdataArray objectAtIndex:self.selectedleftindex]];
        cell.tintColor= [UIColor purpleColor];
        return cell;
    }
   
}


- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 3;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    if (tableView== self.leftTableView) {
        self.selectedleftindex=indexPath.row;
        [self.selectedrighttindexs removeAllObjects];
        [self.rightTableview reloadData];
    }
    if (tableView == self.rightTableview){
        if (![self.selectedrighttindexs containsObject:[NSNumber numberWithInteger:indexPath.row]]) {
                   [self.selectedrighttindexs addObject:[NSNumber numberWithInteger:indexPath.row]];
            NSLog(@"selectedrighttindexs are : %@",self.selectedrighttindexs);
        }    }
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (tableView == self.rightTableview){
     [self.selectedrighttindexs removeObject:[NSNumber numberWithInteger:indexPath.row]];
      NSLog(@"selectedrighttindexs are : %@",self.selectedrighttindexs);
    }
}

@end

File: GullyTrends/ViewControllers/FilterViewController/FilterViewController.swift
//
//  FilterViewController.swift
//  SampleSwift
//
//  Created by Lakshmi Vajrapu on 28/12/19.
//  Copyright © 2019 Lakshmi Vajrapu. All rights reserved.
//

import UIKit
import Parse

class FilterViewController: UIViewController , UITableViewDelegate , UITableViewDataSource{
    
    @IBOutlet var leftTableView: UITableView!
    @IBOutlet var rightTableview: UITableView!
    
    var leftdataArray:Array<String> = [];
    var rightdataArray:Array<PFObject> = [];
    var selectedrighttindexs:NSMutableArray = []
    var preselectedrighttindexs:NSMutableArray = []
    var selectedindexs = [NSInteger:Any]()
    var selectedleftindex:NSInteger = 0
    var filtermodelObjects:FliterModel!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
           //Initial call request Settings pull the all records
           MenuVM.getSettingsLists { (settingsArray, errorMsg) in
             guard errorMsg == nil else {
               AlertUtilities.showAlert(message: errorMsg!) { _ in }
               return
             }
             
             //Filter menu type list from Settings
            FliterVM.getFliterList(array: settingsArray) { (filterarray) in
                print(filterarray)
                self.filtermodelObjects = filterarray
                self.leftdataArray = Array(filterarray.dictionary!.keys)
                self.rightdataArray = filterarray.dictionary![self.leftdataArray[0]]!
                self.leftTableView.reloadData()
                self.rightTableview.reloadData()
                
            }
            
            }
               
           }
           
    override func viewWillAppear(_ animated: Bool) {
        
        leftTableView.register(UINib(nibName: "LeftTableViewCell", bundle: nil), forCellReuseIdentifier: "LeftTableViewCell")
        rightTableview.register(UINib(nibName: "RightTableViewCell", bundle: nil), forCellReuseIdentifier: "RightTableViewCell")
        
        leftTableView.delegate = self
        rightTableview.delegate = self
        
        rightTableview.setEditing(true, animated: true)

    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if tableView == leftTableView {
            return self.leftdataArray.count
        }
      else {
            return rightdataArray.count
        }
    }

    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 50
    }
    
        func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
            if tableView == leftTableView {
                let tableviewcell = tableView.dequeueReusableCell(withIdentifier: "LeftTableViewCell") as! LeftTableViewCell
                tableviewcell.textLabel?.text = leftdataArray[indexPath.row] as? String
                return tableviewcell
            }
            else {
                
                let tableviewcell = tableView.dequeueReusableCell(withIdentifier: "RightTableViewCell") as! RightTableViewCell
                tableviewcell.textLabel?.text = rightdataArray[indexPath.row]["name"] as? String
                tableviewcell.tintColor = UIColor.purple
                return tableviewcell
            }
        
        }
    
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        if (preselectedrighttindexs.contains(indexPath.row) && preselectedrighttindexs.count>0){
                           cell.setSelected(true, animated: true)
                       } else
                       {
                           cell.setSelected(false, animated:false)
                       }
    }
    
    func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        return UITableViewCell.EditingStyle.init(rawValue: 3)!
    }
        
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if tableView == leftTableView {
            
            selectedleftindex = indexPath.row
            rightdataArray = filtermodelObjects.dictionary![leftdataArray[selectedleftindex]]!
            selectedrighttindexs = NSMutableArray()
            preselectedrighttindexs = NSMutableArray()
            if selectedindexs[selectedleftindex] != nil {
                preselectedrighttindexs = selectedindexs[selectedleftindex] as! NSMutableArray
            }
            rightTableview.reloadData()
        }
        if tableView == rightTableview{
        if !(selectedrighttindexs.contains(indexPath.row)){
            selectedrighttindexs.add(indexPath.row)
            selectedindexs[selectedleftindex] = selectedrighttindexs
            
            print(" selectedleftindex:  ",selectedleftindex )
            print(" selectedrighttindexs:  ",selectedrighttindexs )
            print(" selectedindexs:  ",selectedindexs )
            }
           
    }
    }
    func tableView(_ tableView: UITableView,didDeselectRowAt  indexPath: IndexPath) {
        if tableView == rightTableview {
            selectedrighttindexs.remove(indexPath.row)
            print(" selectedrighttindexs:  ",selectedrighttindexs )
        }
    }
        
}
